pipelined/bevy_render2/src/render_component.rs

```diff
@@ -41,7 +41,7 @@ pub trait ExtractComponent: Component {
     /// Filters the entities with additional constraints.
     type Filter: WorldQuery;
     /// Defines how the component is transferred into the "render world".
-    fn extract_component(item: QueryItem<Self::Query>) -> Self;
+    fn extract_component(item: ReadOnlyQueryItem<Self::Query>) -> Self;
```

Comment:
It is late and I can't figure out how to satisfy this bound.

Basically we can't ensure `QueryItem` is returned by `iter` due to needing to allow `Item` to be different to properly support `WriteQuery`... Since it returns `&T` or `Mut<T>` depending on whether you call `iter` or `iter_mut`.

I briefly considered making it work with a singular return type but I don't believe there is a way to do that without a footgun. There isn't a singular type that encapsulates those two types without runtime checks which is a terrible setup.

I believe if `QueryItem` is the way forward the best solve is to keep the idea of two types (`QueryItem` and `ReadOnlyQueryItem`) but name them in a not terrible way.

Reply:
I think the best solution (at least for now) is to just use `iter_mut()` in the ExtractComponent impl and remove the ReadOnlyFetch constraint. It kind of punts this problem forward, but it is niche enough that we might never need to care. And that way we can remove the extra ReadOnlyQueryItem type, which does feel a bit odd. I pushed a commit that does this.

Reply:
The ReadOnlyFetch constraint wasn't needed in this case anyway. I think that was added to accommodate an older design.